lib/services/role.go

```diff
@@ -2828,15 +2828,84 @@ func (set RoleSet) CanForwardAgents() bool {
 	return false
 }
 
-// CanPortForward returns true if a role in the RoleSet allows port forwarding.
-func (set RoleSet) CanPortForward() bool {
+type SSHPortForwardMode int
+
+const (
+	// SSHPortForwardModeOn is the default mode, both remote and local port forwarding is allowed
+	SSHPortForwardModeOn SSHPortForwardMode = iota
+	// SSHPortForwardModeOff disallows any port forwarding.
+	SSHPortForwardModeOff
+	// SSHPortForwardModeRemote allows remote port forwarding.
+	SSHPortForwardModeRemote
+	// SSHPortForwardModeLocal allows local port forwarding.
+	SSHPortForwardModeLocal
+)
+
+// String implements the Stringer interface for SSHPortForwardMode
+func (m SSHPortForwardMode) String() string {
+	switch m {
+	case SSHPortForwardModeOff:
+		return "off"
+	case SSHPortForwardModeLocal:
+		return "local"
+	case SSHPortForwardModeRemote:
+		return "remote"
+	default:
+		return "on"
+	}
+}
+
+// TODO (eriktate): remove Legacy method in v20
+// Legacy returns a bool that can be used as the legacy PortForwarding value when dealing with
+// an agent that doesn't know about SSHPortForwardMode.
+func (m SSHPortForwardMode) Legacy() bool {
+	return m != SSHPortForwardModeOff
+}
+
+// SSHPortForwardMode returns the SSHPortForwardMode permitted by a RoleSet.
+func (set RoleSet) SSHPortForwardMode() SSHPortForwardMode {
+	// port forwarding is allowed by default, we want to track explicit denies
+	allowRemote := false
+	allowLocal := false
+
 	for _, role := range set {
```

Comment:
suggestion:
- only use the legacy behavior if no role in the entire roleset uses the new options
- if using the new behavior, a single explicit legacy deny is a total deny
- a single explicit local or remote deny in the new option should result in that local/remote type being denied

something like this:
```go
var (
	legacyDeny bool
	legacyAllow
	denyLocal
	denyRemote
	usingNewOptions
)
for _, role := range set {
	legacyDeny |= !types.BoolDefaultTrue(role.GetOptions().PortForwarding)
	legacyAllow |= types.BoolDefaultTrue(role.GetOptions().PortForwarding)	
	config := role.GetOptions().SSHPortForwarding
	if config == nil {
		continue
	}
	usingNewOptions = true
	if config.Remote != nil {
		denyRemote |= !types.BoolDefaultTrue(config.Remote.Enabled)
	}
	if config.Local != nil {
		denyLocal |= !types.BoolDefaultTrue(config.Local.Enabled)
	}
}
if !usingNewOptions {
	// Preserve legacy behavior, a single allow or unset option allows portforwarding
	if legacyAllow {
		return SSHPortForwardModeOn
	}
	return SSHPortForwardModeOff
}
switch {
case legacyDeny || denyLocal && denyRemote:
	return SSHPortForwardModeOff
case denyLocal:
	return SSHPortForwardModeRemote
case denyRemote:
	return SSHPortForwardModeLocal
default:
	return SSHPortForwardModeOn
}
```

Reply:
> That's the opposite of how port_forwarding works, though

Good, the way it currently works is terrible, I see this as an opportunity for us to rectify our past mistakes 

> it seems like a potential land mine if adding ssh_port_forwarding to a single role effectively invalidates other roles that appear like they should allow forwarding

imo it's a far bigger landmine if you can have an existing role with `ssh_port_forwarding: {server: false}`, and then adding another role that doesn't mention port forwarding at all suddenly allows server port forwarding

Reply:
But yes we should definitely avoid breaking changes, that's why I think using the new field should be an opt-in to the new behavior, nothing will change until people start using this

Looks like you were discussing this with @rosstimothy, I'd be happy to chat about this with both of you, i feel fairly strongly we should take this opportunity to avoid the current state where roles that don't even mention port forwarding can completely override apparent "denies" in other roles

I am open to dropping my second bullet though "if using the new field, a single explicit legacy deny is a total deny", don't really care about this as long as behavior with roles that use the new field or nothing is sane

Reply:
@nklaassen what if we biased towards respecting the legacy behavior while preferring explicit denies in the new `ssh_port_forwarding` config? I just pushed a commit that should represent what I mean exactly, but the most important points are:
- `port_forwarding` can be left undefined now rather than always defaulting to true.
- If a role has an `ssh_port_forwarding` configured, the `port_forwarding` field on the same role is ignored.
- If a role has `port_forwarding: true` defined with no `ssh_port_forwarding` config, that's an immediate allow all which is what we would expect to happen today.
- If no roles are using the legacy `port_forwarding` field, `ssh_port_forwarding` biases towards explicit denies.

I think this maintains full backwards compatibility with existing behavior without forcing `ssh_port_forwarding` to follow the same semantics. It also makes it possible for users to opt fully into the new model as long as they remove `port_forwarding` from all of their existing roles.

Reply:
> If a role has an ssh_port_forwarding configured, the port_forwarding field on the same role is ignored.

Can we make it an error to create or update a role with both set?